client.go

```diff
@@ -920,6 +952,91 @@ func (c *Client) GenerateClientAssertion() (assertion string, err error) {
 	return t.SignedString(c.clientAssertionSigningKey)
 }
 
+type PARResponse struct {
+	ExpiresIn  int64  `json:"expires_in,omitempty"`
+	RequestURI string `json:"request_uri,omitempty"`
+}
+
+func (c *Client) DoPAR(options ...AuthorizeOption) (pr PARResponse, csrf CSRF, err error) {
+	var (
+		body     []byte
+		request  *http.Request
+		response *http.Response
+	)
+
+	if csrf.State, err = randomString(StateLength); err != nil {
+		return pr, csrf, err
+	}
+
+	if csrf.Nonce, err = randomString(NonceLength); err != nil {
+		return pr, csrf, err
+	}
+
+	values := url.Values{
+		"client_id": {c.Config.ClientID},
+		"nonce":     {csrf.Nonce},
+		"state":     {csrf.State},
+	}
+
+	if c.Config.RedirectURL != nil {
+		values.Set("redirect_uri", c.Config.RedirectURL.String())
+	}
+
+	if len(c.Config.Scopes) > 0 {
+		values.Set("scope", strings.Join(c.Config.Scopes, " "))
+	}
+
+	if c.Config.AuthMethod == ClientSecretPostAuthnMethod {
+		values.Add("client_secret", c.Config.ClientSecret)
+	}
+
+	if c.Config.AuthMethod == PrivateKeyJwtAuthnMethod {
+		var (
+			assertion string
+		)
+		values.Add("client_assertion_type", "urn:ietf:params:oauth:client-assertion-type:jwt-bearer")
+		if assertion, err = c.GenerateClientAssertion(); err != nil {
+			return pr, csrf, errors.Wrapf(err, "failed to generate client assertion")
+		}
+		values.Add("client_assertion", assertion)
+	}
+
+	for _, o := range options {
+		if err = o.apply(c, values, &csrf); err != nil {
+			return pr, csrf, err
+		}
+	}
```

Comment:
Is the logic responsible for preparing the request (url.Values) the same as in the Authorize method?
Maybe it could be simplified.

Reply:
I merged all code  which was duplicated. Please review. It is still not tested with openbanking-quickstart but should work.